chebi-update/src/main/java/org/reactome/release/chebiupdate/ChebiUpdater.java

```diff
@@ -0,0 +1,487 @@
+	private static final Logger logger = LogManager.getLogger("ChEBIUpdateLogger");
+	private boolean testMode = true;
+	private MySQLAdaptor adaptor;
+	private ChebiWebServiceClient chebiClient = new ChebiWebServiceClient();
+	private StringBuilder identifierSB = new StringBuilder();
+	private StringBuilder formulaUpdateSB = new StringBuilder();
+	private StringBuilder formulaFillSB = new StringBuilder();
+	private StringBuilder nameSB = new StringBuilder();
+	private StringBuilder duplicatesSB = new StringBuilder();
+
+	private long personID;
+	
+	public ChebiUpdater(MySQLAdaptor adaptor, boolean testMode, long personID)
+	{
+		this.adaptor = adaptor;
+		this.testMode = testMode;
+		this.personID = personID;
+	}
+
+	/**
+	 * Update ChEBI ReferenceMolecules. This method will query ChEBI for up-to-date
+	 * information, and using that information it will: <br/>
+	 * <ul>
+	 * <li>Update the names of ReferenceEntitites that are refer to the
+	 * ReferenceMolecule</li>
+	 * <li>Update the names of ReferenceMolecules</li>
+	 * <li>Update the identifiers of ReferenceMolecules</li>
+	 * <li>Update the formulae of ReferenceMolecules</li>
+	 * </ul>
+	 * 
+	 * @throws SQLException
+	 * @throws Exception
+	 */
+	public void updateChebiReferenceMolecules() throws SQLException, Exception
+	{
+		@SuppressWarnings("unchecked")
+		String chebiRefDBID = (new ArrayList<GKInstance>( adaptor.fetchInstanceByAttribute("ReferenceDatabase", "name", "=", "ChEBI"))).get(0).getDBID().toString();
+
+		@SuppressWarnings("unchecked")
+		Collection<GKInstance> refMolecules = (Collection<GKInstance>) adaptor.fetchInstanceByAttribute("ReferenceMolecule", "referenceDatabase", "=", chebiRefDBID);
+
+		// A map: key is the DB_ID of a ReferneceMolecule, value is the
+		// uk.ac.ebi.chebi.webapps.chebiWS.model.Entity from ChEBI.
+		Map<Long, Entity> entityMap = Collections.synchronizedMap(new HashMap<Long, Entity>());
+		// A list of the ReferenceMolecules where we could nto get info from ChEBI.
+		List<GKInstance> failedEntiesList = Collections.synchronizedList(new ArrayList<GKInstance>());
+
+		logger.info("{} ChEBI ReferenceMolecules to check...", refMolecules.size());
+
+		retrieveUpdatesFromChebi(refMolecules, entityMap, failedEntiesList);
+
+		logger.info("Number of entities we were able to retrieve information about: {}", entityMap.size());
+		logger.info("Number of entities we were NOT able to retrieve information about: {}", failedEntiesList.size());
+
+		for (GKInstance molecule : failedEntiesList)
+		{
+			logger.info("Could not get info from ChEBI for: {}", molecule.toString());
+		}
+		
+		GKInstance instanceEdit = null;
+		if (!testMode)
+		{
+			instanceEdit = createInstanceEdit(this.adaptor, this.personID, this.getClass().getCanonicalName());
+		}
+		
+		for (Long moleculeDBID : entityMap.keySet())
+		{
+			// One transaction per molecule - is this too many? If this runs too slow, maybe switch to one transaction per
+			// program execution.
```

Comment:
Guanming did recommend trying to minimize DB interactions, is there any specific benefit to having one transaction per molecule vs doing a bulk update?

Reply:
Transactions are supposed to be atomic - if it succeeds, the whole thing succeeds. If it fails, the whole thing fails. In this case, I considered that if one transaction fails (for some reason), I didn't want ALL of the data update transactions to fail. 
Also, this program only takes a couple of minutes to run so I don't think the performance overhead is too bad. I could change it to one large transaction. If that fails, all data updates would need to be redone. In a program of this scale, it probably doesn't matter which way it's done. I could do timing tests later (if i have time) to see if multiple transactions is significantly slower or not.

Reply:
Can't argue with that logic -- better to fail a tiny bit for a good reason then to fail everything due to a tiny reason. Also, the module is down to minutes? Now thats performance improvement!

Reply:
On my workstation, a few minutes at most.